File: mobile/ios/OpenU60/Features/RouterSettings/APN/APNViewModel.swift
import SwiftUI

@Observable
@MainActor
final class APNViewModel {
    var config: APNConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false
    var showFormSheet: Bool = false

    // Form state
    var formProfile: APNProfile = .empty
    var editingProfile: APNProfile?  // nil = adding, non-nil = editing
    var setAsDefault: Bool = false

    private let client: AgentClient
    private let authManager: AuthManager

    var isEditing: Bool { editingProfile != nil }

    var activeAPNName: String? {
        let active = config.profiles.first(where: { $0.active })
            ?? config.autoProfiles.first(where: { $0.active })
        guard let active else { return nil }
        return active.name.isEmpty ? active.apn : active.name
    }

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let modeData = try await client.getJSON("/api/router/apn/mode")
            let mode = APNParser.parseMode(modeData)

            let manuData = try await client.getJSON("/api/router/apn/profiles")
            let profiles = APNParser.parseProfiles(manuData)

            let autoData = try await client.getJSON("/api/router/apn/auto-profiles")
            let autoProfiles = APNParser.parseProfiles(autoData)

            config = APNConfig(mode: mode, profiles: profiles, autoProfiles: autoProfiles)
        } catch {
            showMessage("Failed to load APN: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func setMode(manual: Bool) async {
        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/apn/mode", body: ["apn_mode": manual ? 1 : 0])
            showMessage("APN mode set to \(manual ? "manual" : "auto")", isError: false)
            config = APNConfig(
                mode: manual ? "1" : "0",
                profiles: config.profiles,
                autoProfiles: config.autoProfiles
            )
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    // MARK: - Form Actions

    func startAdd() {
        editingProfile = nil
        formProfile = .empty
        setAsDefault = false
        showFormSheet = true
    }

    func startEdit(_ profile: APNProfile) {
        editingProfile = profile
        formProfile = profile
        setAsDefault = profile.active
        showFormSheet = true
    }

    func saveAPN() async {
        guard !formProfile.name.isEmpty, !formProfile.apn.isEmpty else {
            showMessage("Name and APN are required", isError: true)
            return
        }

        let isDuplicate = config.profiles.contains { p in
            p.name == formProfile.name && p.id != (editingProfile?.id ?? "")
        }
        if isDuplicate {
            showMessage("An APN with this name already exists", isError: true)
            return
        }

        if isEditing {
            await editAPN()
        } else {
            await addAPN()
        }
    }

    private func addAPN() async {
        isLoading = true

        let name = formProfile.name
        let apn = formProfile.apn
        let shouldSetDefault = setAsDefault

        do {
            let _ = try await client.postJSON("/api/router/apn/profiles", body: [
                "profilename": name,
                "wanapn": apn,
                "pdpType": formProfile.pdpType,
                "pppAuthMode": formProfile.authMode,
                "username": formProfile.username,
                "password": formProfile.password
            ])

            if shouldSetDefault {
                await refresh()
                if let newProfile = config.profiles.first(where: { $0.name == name && $0.apn == apn }) {
                    await activateAPN(newProfile)
                }
            }

            formProfile = .empty
            showFormSheet = false
            showMessage("APN added", isError: false)
            await refresh()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func editAPN() async {
        guard let editing = editingProfile else { return }

        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/apn/profiles", body: [
                "profileId": editing.id,
                "profilename": formProfile.name,
                "wanapn": formProfile.apn,
                "pdpType": formProfile.pdpType,
                "pppAuthMode": formProfile.authMode,
                "username": formProfile.username,
                "password": formProfile.password
            ])

            if setAsDefault && !editing.active {
                await activateAPN(editing)
            }

            editingProfile = nil
            showFormSheet = false
            showMessage("APN updated", isError: false)
            await refresh()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func deleteAPN(_ profile: APNProfile) async {
        if profile.active {
            showMessage("Cannot delete the active APN", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.postJSON("/api/router/apn/profiles/delete", body: ["profileId": profile.id])
            showMessage("APN deleted", isError: false)
            config.profiles.removeAll { $0.id == profile.id }
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func activateAPN(_ profile: APNProfile) async {
        isLoading = true

        do {
            let _ = try await client.postJSON("/api/router/apn/profiles/activate", body: ["profileId": profile.id])
            showMessage("APN activated", isError: false)
            config.profiles = config.profiles.map { p in
                var updated = p
                updated.active = p.id == profile.id
                return updated
            }
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/apn/APNViewModel.kt
package com.openu60.feature.router.apn

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.APNConfig
import com.openu60.core.model.APNParser
import com.openu60.core.model.APNProfile
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class APNState(
    val config: APNConfig = APNConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
    val showForm: Boolean = false,
    val editingProfile: APNProfile? = null,
)

@HiltViewModel
class APNViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(APNState())
    val state: StateFlow<APNState> = _state.asStateFlow()

    fun showAddForm() {
        _state.value = _state.value.copy(showForm = true, editingProfile = APNProfile.empty)
    }

    fun showEditForm(profile: APNProfile) {
        _state.value = _state.value.copy(showForm = true, editingProfile = profile)
    }

    fun hideForm() {
        _state.value = _state.value.copy(showForm = false, editingProfile = null)
    }

    fun updateEditingProfile(profile: APNProfile) {
        _state.value = _state.value.copy(editingProfile = profile)
    }

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/modem/apn")
                val mode = APNParser.parseMode(data)
                val profiles = APNParser.parseProfiles(data)
                _state.value = _state.value.copy(
                    config = APNConfig(mode = mode, profiles = profiles),
                    isLoading = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun setMode(mode: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/modem/apn/mode", mapOf("apn_mode" to mode))
                _state.value = _state.value.copy(
                    config = _state.value.config.copy(mode = mode),
                    isLoading = false,
                    message = if (mode == "1") "Manual APN mode" else "Auto APN mode",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) setMode(mode) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun saveProfile() {
        val profile = _state.value.editingProfile ?: return
        val existingNames = _state.value.config.profiles
            .filter { it.id != profile.id }
            .map { it.name.lowercase() }
        if (profile.name.lowercase() in existingNames) {
            _state.value = _state.value.copy(message = "Profile name already exists", messageIsError = true)
            return
        }

        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val params = mapOf(
                    "profileId" to profile.id,
                    "profilename" to profile.name,
                    "wanapn" to profile.apn,
                    "pdpType" to profile.pdpType,
                    "pppAuthMode" to profile.authMode,
                    "username" to profile.username,
                    "password" to profile.password,
                )
                if (profile.id.isNotBlank()) {
                    agentClient.putJSON("/api/modem/apn/profile", params)
                } else {
                    agentClient.postJSON("/api/modem/apn/profile", params)
                }
                _state.value = _state.value.copy(showForm = false, editingProfile = null, isLoading = false, message = "Profile saved", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) saveProfile() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun deleteProfile(id: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.deleteJSON("/api/modem/apn/profile", mapOf("profileId" to id))
                _state.value = _state.value.copy(isLoading = false, message = "Profile deleted", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) deleteProfile(id) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun activateProfile(id: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/modem/apn/activate", mapOf("profileId" to id))
                _state.value = _state.value.copy(isLoading = false, message = "Profile activated", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) activateProfile(id) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}
